src/TagInputStateful/index.jsx

```diff
@@ -0,0 +1,623 @@
+                ? getOption( optionId, flatOptions ).id : undefined;
+        }
+
+        return {
+            filteredOptions : state.filteredOptions || props.options,
+            id              : props.id || state.id || generateId( 'Select' ),
+            isOpen          : state.isOpen,
+            options         : props.options,
+            value           : state.value,
+            selection       : optionId,
+        };
+    }
+
+    handleBlur( e )
+    {
+        const callback = this.props.onBlur;
+
+        if ( callback )
+        {
+            callback( e );
+        }
+
+        this.setState( { isOpen: false } );
+    }
+
+    handleChange( e )
+    {
+        const callback = this.props.onChange;
+
+        if ( callback )
+        {
+            callback( e );
+        }
+
+        const searchValue = ( e.target.value || '' ).toLowerCase();
+
+        const filteredOptions = this.props.options.filter( ( { text } ) =>
+            !searchValue || text.toLowerCase().indexOf( searchValue ) > -1 );
+
+        const activeOption = ( searchValue && filteredOptions.length ) ?
+            filteredOptions[ 0 ].id : undefined;
+
+        this.setState( {
+            activeOption,
+            filteredOptions,
+            value : e.target.value,
+        } );
+    }
+
+    handleClickClose( e )
+    {
+        const { tags } = this.state;
+        const callback = this.props.onClickClose;
+        const closeTag = e.target.value;
+
+        if ( callback )
+        {
+            callback( closeTag );
+        }
+
+        const newItems = tags.filter( ( item ) => ( typeof item === 'string' ?
+            item !== closeTag : item.value !== closeTag ) );
+
+        this.setState( { tags: newItems } );
+    }
+
+    handleClickOption( e, id )
+    {
+        const { tags } = this.state;
+        const callback = this.props.onClickOption;
+
+        if ( callback )
+        {
+            callback( e );
+        }
+
+        const option = this.props.options.find( opt => opt.id === id );
+        const newTags = [ ...tags, option.text ];
+
+        this.setState( prevState =>
+        {
+            const selectedOption = getOption( id, prevState.options );
+            return {
+                activeOption : selectedOption.id,
+                selection    : selectedOption.id,
+                value        : undefined,
+                tags         : newTags,
+            };
+        } );
+    }
+
+    handleFocus( e )
+    {
+        const callback = this.props.onFocus;
+
+        if ( callback )
+        {
+            callback( e );
+        }
+
+        this.setState( { isOpen: true } );
+    }
+
+    handleKeyDown( e )
+    {
+        const BACKSPACE = 8;  // keyCode for 'backspace'
+        const ENTER     = 13; // keyCode for 'enter'
+        const ARROWUP   = 38; // keyCode for 'arrow up'
+        const ARROWDOWN = 40; // keyCode for 'arrow down'
+        const ESCAPE    = 27; // keyCode for 'escape'
```

Comment:
Instead of using `keycode` we can just use `key` and make things easier for ourselves 😄 (`key` already returns the name of the key pressed, e.g: the string `'Enter'` rather than the number `13`).

---
Also, minor side note: if we _did_ need these constants then they should be declared at the top of the file rather than re-initializing them every time `handleKeyDown` is called (`const` and `let` do not get hoisted).